packages/editor/src/components/post-publish-panel/media-util.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import { v4 as uuid } from 'uuid';
+
+/**
+ * WordPress dependencies
+ */
+import { getFilename } from '@wordpress/url';
+
+// Generate a list of unique filenames given a list of URLs.
+// We want all basenames to be unique, since sometimes the extension
+// doesn't reflect the mime type, and may end up getting changed by
+// the server, on upload.
+export function generateUniqueFilenames( urls ) {
+	const basenames = new Set();
+
+	return urls.map( ( url ) => {
+		// We prefer to match the remote filename, if possible.
+		const filename = getFilename( url );
+		let basename = '';
+		let extension = '';
+
+		if ( filename ) {
+			const parts = filename.split( '.' );
+			if ( parts.length > 1 ) {
+				// Assume the last part is the extension.
+				extension = parts.pop();
+			}
+			basename = parts.join( '.' );
+		}
+
+		if ( ! basename ) {
+			// It looks like we don't have a basename, so let's use a UUID.
+			basename = uuid();
+		}
+
+		if ( basenames.has( basename ) ) {
+			// Append a UUID to deduplicate the basename.
+			// The server will try to deduplicate on its own if we don't do this,
+			// but it may run into a race condition
+			// (see https://github.com/WordPress/gutenberg/issues/64899).
+			// Deduplicating the filenames before uploading is safer.
+			basename = `${ basename }-${ uuid() }`;
+		}
+
+		basenames.add( basename );
+
+		return { url, basename, extension };
+	} );
+}
+
+const mimeTypeToExtension = {
+	'image/apng': 'apng',
+	'image/avif': 'avif',
+	'image/bmp': 'bmp',
+	'image/gif': 'gif',
+	'image/vnd.microsoft.icon': 'ico',
+	'image/jpeg': 'jpg',
+	'image/jxl': 'jxl',
+	'image/png': 'png',
+	'image/svg+xml': 'svg',
+	'image/tiff': 'tiff',
+	'image/webp': 'webp',
+	'video/x-msvideo': 'avi',
+	'video/mp4': 'mp4',
+	'video/mpeg': 'mpeg',
+	'video/ogg': 'ogg',
+	'video/mp2t': 'ts',
+	'video/webm': 'webm',
+	'video/3gpp': '3gp',
+	'video/3gpp2': '3g2',
+};
+
+// Get the file extension to use for a given mime type.
+export function getExtensionFromMimeType( mime ) {
+	mime = ( mime ?? '' ).toLowerCase();
+
+	let extension = mimeTypeToExtension[ mime ];
+
+	if ( ! extension ) {
+		// We don't know which extension to use, so we need to fall back to
+		// something safe. The server should replace it with an appropriate
+		// extension for the mime type.
+		if ( mime.startsWith( 'image/' ) ) {
+			extension = 'png';
+		}
+		if ( mime.startsWith( 'video/' ) ) {
+			extension = 'mp4';
+		}
+	}
+
+	// If all else fails, try an empty extension.
+	// The server will probably reject the upload, but there isn't much
+	// else we can do.
+	return extension || '';
+}
```

Comment:
This may well be overkill, and perhaps some simplification is in order 👍 I think your suggestion may be going a bit too far in the opposite direction, though.

Wouldn't it be best for us to code a bit defensively here? I'm not sure I like the scenario where the user adds a plugin to enable SVG support and ends up trying to upload `image.svg+xml` 😕

Reply:
I almost included that in my suggestion, but yes we could add a hardcoded return value for SVGs specifically. It's really an edge case though :-)


```js
export function getExtensionFromMimeType( mime ) {
	if ( 'image/svg+xml' === mimeType ) {
		return 'svg';
	}

	return mimeType.split( '/' )[ 1 ] || '';
}
```

In my media experiments work I also used [`mime/lite`](https://www.npmjs.com/package/mime) in the past, which is more accurate, but IMO that's also overkill here.

Reply:
We went with a fixed extension instead, and are relying on the server to apply the correct one (see other thread).